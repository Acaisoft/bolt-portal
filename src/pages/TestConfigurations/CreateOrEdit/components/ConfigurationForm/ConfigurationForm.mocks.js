/**
 * Copyright (c) 2022 Acaisoft
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { mockGraphqlData } from 'utils/tests/mocks'
import {
  GET_CONFIGURATION,
  GET_CONFIGURATION_TYPES,
  GET_PARAMETERS,
  GET_TEST_SOURCES_FOR_PROJECT,
} from './graphql'

const mockedConfigurationTypesData = {
  configurationTypes: [
    {
      id: 'fc9c1146-44d7-4054-a12c-4c0f70487230',
      name: 'Performance',
      slug_name: 'load_tests',
    },
  ],
}

const testSourceId = '529606f0-58ca-4ee6-8374-2bf4ca6d5635'
const mockedTestSourcesData = {
  testSources: [
    {
      id: testSourceId,
      source_type: 'repository',
      repository: {
        id: 'd8d3c51d-d177-4bf3-bde9-0309aa062701',
        name: 'LoadTestsRepo',
        type_slug: 'load_tests',
      },
      test_creator: null,
    },
  ],
}

const testParametersMockData = {
  parameters: [
    {
      id: '9352e6d1-229a-4938-9e9f-9ea8ab0600e0',
      name: 'time',
      param_name: '-t',
      param_type: 'str',
      default_value: '10',
      slug_name: 'load_tests_duration',
      type_slug: 'load_tests',
    },
    {
      id: '35f63de7-b372-4eb2-8f67-a471a6c67b50',
      name: 'users/second',
      param_name: '-r',
      param_type: 'int',
      default_value: '500',
      slug_name: 'load_tests_rampup',
      type_slug: 'load_tests',
    },
    {
      id: '7dfe8616-365a-4aec-895f-f8a7eb8f30c6',
      name: 'users',
      param_name: '-c',
      param_type: 'int',
      default_value: '1000',
      slug_name: 'load_tests_users',
      type_slug: 'load_tests',
    },
    {
      id: '7a8423d3-88ce-4277-ac8d-131a6fb32314',
      name: 'host',
      param_name: '-H',
      param_type: 'str',
      default_value: '',
      slug_name: 'load_tests_host',
      type_slug: 'load_tests',
    },
    {
      id: '97285024-0d4d-439d-abee-705c04cafec8',
      name: 'monitoring duration',
      param_name: '-md',
      param_type: 'int',
      default_value: '10',
      slug_name: 'monitoring_duration',
      type_slug: 'load_tests',
    },
    {
      id: '8a8ef988-62b6-4338-a350-9712e00b024b',
      name: 'monitoring interval',
      param_name: '-mi',
      param_type: 'int',
      default_value: '5',
      slug_name: 'monitoring_interval',
      type_slug: 'load_tests',
    },
    {
      id: 'c5c537f7-e1a8-40e7-ac50-45f54dc592bd',
      name: 'file name',
      param_name: '-f',
      param_type: 'str',
      default_value: 'load_tests.py',
      slug_name: 'load_tests_file_name',
      type_slug: 'load_tests',
    },
    {
      id: 'a426ff2e-b12c-49e9-b0ef-79da32d6a113',
      name: 'repository branch',
      param_name: '-b',
      param_type: 'str',
      default_value: 'master',
      slug_name: 'load_tests_repository_branch',
      type_slug: 'load_tests',
    },
  ],
}

export const projectId = 'f3e2db7e-20b7-43c7-a32c-a47f958a2647'
export const configurationId = '6f0428ac-0579-4a65-bb86-2bb049398b57'
export const testConfigurationBase = {
  configuration: {
    id: configurationId,
    name: 'google (Cloned at 22/04/2022 - 09:30:30) MB',
    configuration_parameters: [
      {
        parameter_slug: 'load_tests_duration',
        value: '20',
      },
      {
        parameter_slug: 'load_tests_rampup',
        value: '1',
      },
      {
        parameter_slug: 'load_tests_users',
        value: '5',
      },
      {
        parameter_slug: 'load_tests_host',
        value: 'http://google.com',
      },
      {
        parameter_slug: 'load_tests_file_name',
        value: 'load_tests.py',
      },
      {
        parameter_slug: 'load_tests_repository_branch',
        value: 'master',
      },
    ],
    type_slug: 'load_tests',
    test_source: {
      id: testSourceId,
      source_type: 'repository',
    },
    has_pre_test: false,
    has_post_test: false,
    has_load_tests: true,
    has_monitoring: false,
    configuration_envvars: [
      {
        name: 'TEST_KEY',
        value: 'TEST_VALUE',
      },
      {
        name: 'NEW_SECRET_KEY',
        value: 'SECRET_VALUE',
      },
    ],
  },
}

export function getParameterLabel(slug) {
  const item = testParametersMockData.parameters.find(
    ({ slug_name }) => slug_name === slug
  )

  if (!item) return ''
  return item.name
}

const testConfigurationPerformed = {
  configuration: {
    ...testConfigurationBase.configuration,
    performed: true,
  },
}

const testConfigurationNotPerformed = {
  configuration: {
    ...testConfigurationBase.configuration,
    performed: false,
  },
}

export const configurationTypesMock = mockGraphqlData(
  GET_CONFIGURATION_TYPES,
  mockedConfigurationTypesData
)

export const testSourcesMock = mockGraphqlData(
  GET_TEST_SOURCES_FOR_PROJECT,
  mockedTestSourcesData,
  { projectId }
)

export const testParametersMock = mockGraphqlData(
  GET_PARAMETERS,
  testParametersMockData
)

export const testConfigurationPerformedMock = mockGraphqlData(
  GET_CONFIGURATION,
  testConfigurationPerformed,
  { configurationId }
)
export const testConfigurationNotPerformedMock = mockGraphqlData(
  GET_CONFIGURATION,
  testConfigurationNotPerformed,
  { configurationId }
)
