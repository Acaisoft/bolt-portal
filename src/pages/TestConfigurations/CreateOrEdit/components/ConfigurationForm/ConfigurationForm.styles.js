import { makeStyles } from '@material-ui/core'

export default makeStyles(({ spacing }) => ({
  button: {
    marginLeft: spacing(2),
  },
}))
