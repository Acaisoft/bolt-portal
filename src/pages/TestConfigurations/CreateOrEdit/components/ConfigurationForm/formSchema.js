import { useMemo } from 'react'
import { useQuery } from 'react-apollo-hooks'

import { TestSourceType } from '~config/constants'
import { validateOnFieldValue } from '~utils/forms'

import {
  GET_PARAMETERS,
  GET_TEST_SOURCES_FOR_PROJECT,
  GET_CONFIGURATION_TYPES,
} from './graphql'

const scenarioParts = [
  {
    id: 'has_pre_test',
    label: 'Before Scenario',
    description: 'before description',
  },
  {
    id: 'has_post_test',
    label: 'After Scenario',
    description: 'after description',
  },
  {
    id: 'has_load_tests',
    label: 'Load Tests',
    description: 'load tests description',
  },
  {
    id: 'has_monitoring',
    label: 'Monitoring',
    description: 'monitoring description',
  },
]

function useFormSchema({ projectId, mode }) {
  const {
    data: { parameters },
    loading: parametersLoading,
  } = useQuery(GET_PARAMETERS, {
    fetchPolicy: 'cache-and-network',
  })
  const {
    data: { configurationTypes },
    loading: configurationTypesLoading,
  } = useQuery(GET_CONFIGURATION_TYPES, {
    fetchPolicy: 'cache-and-network',
  })
  const {
    data: { testSources },
    loading: testSourcesLoading,
  } = useQuery(GET_TEST_SOURCES_FOR_PROJECT, {
    fetchPolicy: 'cache-and-network',
    variables: { projectId },
  })

  const fields = useMemo(
    () =>
      generateFields({
        parameters: parameters || [],
        configurationTypes: configurationTypes || [],
        testSources: testSources || [],
        testSourceTypes: [
          { slug_name: '', label: 'No test source' },
          { slug_name: TestSourceType.REPOSITORY, label: 'Repository' },
          // { slug_name: TestSourceType.TEST_CREATOR, label: 'Test Creator' }, // Disabled for now
        ],
      }),
    [parameters, configurationTypes, testSources]
  )

  return {
    loading: parametersLoading || configurationTypesLoading || testSourcesLoading,
    fields,
  }
}

function generateFields({
  configurationTypes,
  parameters,
  testSources,
  testSourceTypes,
}) {
  const configurationTypeOptions = configurationTypes.map(ct => ({
    key: ct.id,
    label: ct.name,
    value: ct.slug_name,
  }))

  const testSourceTypeOptions = testSourceTypes.map(tst => ({
    key: tst.slug_name,
    label: tst.label,
    value: tst.slug_name,
  }))

  const testSourceOptions = testSources.map(ts => ({
    configuration_type: ts[ts.source_type].type_slug,
    key: ts.id,
    label: ts[ts.source_type].name,
    value: ts.id,
    source_type: ts.source_type,
  }))

  const paramTypeValidators = {
    int: { numericality: { onlyInteger: true } },
    str: { length: { minimum: 1 } },
  }

  const fields = {
    scenario_name: {
      validator: {
        presence: { allowEmpty: false },
      },
      inputProps: {
        label: 'Name',
      },
    },
    configuration_type: {
      validator: {
        inclusion: configurationTypeOptions.map(cto => cto.value),
      },
      options: configurationTypeOptions,
      inputProps: {
        select: true,
        label: 'Test Type',
      },
    },
    scenario_parts: {
      fields: scenarioParts.reduce(
        (acc, sp) => ({
          ...acc,
          [sp.id]: {
            inputProps: {
              label: sp.label,
              helperText: sp.description,
            },
            defaultValue: false,
          },
        }),
        {}
      ),
    },
    parameters: {
      fields: parameters.reduce(
        (acc, parameter) => ({
          ...acc,
          [parameter.slug_name]: {
            validator: (value, formValues) => {
              if (
                formValues.configuration_type === parameter.type_slug &&
                ((formValues.scenario_parts.has_load_tests &&
                  parameter.slug_name.includes('load_tests')) ||
                  (formValues.scenario_parts.has_monitoring &&
                    parameter.slug_name.includes('monitoring')))
              ) {
                return {
                  presence: { allowEmpty: true },
                  ...paramTypeValidators[parameter.param_type],
                }
              }
            },
            inputProps: {
              label: parameter.name,
            },
            defaultValue: parameter.default_value,
            group: parameter.type_slug,
            // TODO: Probably will need to be rewritten
            // when we decide how to store monitoring parameters
            scenarioPart: parameter.slug_name.includes('monitoring')
              ? 'monitoring'
              : 'load_tests',
          },
        }),
        {}
      ),
    },
    test_source_type: {
      validator: {
        inclusion: testSourceTypes.map(tst => tst.slug_name),
      },
      options: testSourceTypeOptions,
      inputProps: {
        select: true,
        label: 'Source Type',
      },
    },
    test_source: {
      fields: testSourceTypes.reduce(
        (acc, type) => ({
          ...acc,
          // type.slug_name: repository, test_creator
          [type.slug_name]: {
            validator: validateOnFieldValue('test_source_type', type.slug_name, {
              presence: { allowEmpty: false },
              inclusion: testSources.map(ts => ts.id),
            }),
            options: testSourceOptions.filter(
              tso => tso.source_type === type.slug_name
            ),
            inputProps: {
              select: true,
              label: `Select ${type.label}`,
            },
          },
        }),
        {}
      ),
    },
  }
  return fields
}

function prepareInitialValues(data) {
  if (!data) {
    return {
      configuration_type: 'load_tests',
      configuration_envvars: [{ name: '', value: '' }],
    }
  }

  const {
    name,
    type_slug,
    configuration_parameters,
    performed,
    test_source,
    has_pre_test,
    has_post_test,
    has_load_tests,
    has_monitoring,
    configuration_envvars,
  } = data

  return {
    scenario_name: name,
    configuration_type: type_slug,
    performed,
    scenario_parts: {
      has_pre_test,
      has_post_test,
      has_load_tests,
      has_monitoring,
    },
    parameters: configuration_parameters.reduce(
      (acc, parameter) => ({
        ...acc,
        [parameter.parameter_slug]: parameter.value,
      }),
      {}
    ),
    test_source_type: test_source && test_source.source_type,
    test_source: test_source ? { [test_source.source_type]: test_source.id } : null,
    configuration_envvars: configuration_envvars.map(({ name, value }) => ({
      name,
      value,
    })) || [{ name: '', value: '' }],
  }
}

function preparePayload(formValues, { mode, configurationId, projectId }) {
  if (!formValues) {
    return {}
  }

  const {
    scenario_name,
    configuration_type,
    scenario_parts: { has_pre_test, has_post_test, has_load_tests, has_monitoring },
    parameters,
    test_source_type,
    test_source,
    configuration_envvars,
  } = formValues

  const variables = {
    name: scenario_name,
    configuration_envvars: configuration_envvars.filter(
      ce => ce.name !== '' && typeof ce.name !== 'undefined'
    ),
  }

  if (mode === 'create') {
    variables.project_id = projectId
  } else {
    variables.id = configurationId
  }

  Object.assign(variables, {
    type_slug: configuration_type,
    has_pre_test,
    has_post_test,
    has_load_tests,
    has_monitoring,
    configuration_parameters: Object.entries(parameters)
      .map(([slug, value]) => ({
        parameter_slug: slug,
        value,
      }))
      // Skip parameters for not checked scenario parts
      .filter(
        ({ parameter_slug }) =>
          (has_load_tests && parameter_slug.includes('load_tests')) ||
          (has_monitoring && parameter_slug.includes('monitoring'))
      ),
    test_source_id: test_source[test_source_type],
  })

  return variables
}

export { useFormSchema, prepareInitialValues, preparePayload }
