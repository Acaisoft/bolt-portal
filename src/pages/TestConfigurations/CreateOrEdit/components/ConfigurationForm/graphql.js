import gql from 'graphql-tag'

export const GET_CONFIGURATION = gql`
  query getTestConfiguration($configurationId: uuid!) {
    configuration: configuration_by_pk(id: $configurationId) {
      id
      name
      performed
      configuration_parameters {
        parameter_slug
        value
      }
      type_slug
      test_source {
        id
        source_type
      }
      has_pre_test
      has_post_test
      has_load_tests
      has_monitoring
      configuration_envvars {
        name
        value
      }
    }
  }
`

export const ADD_CONFIGURATION_MUTATION = gql`
  mutation addTestConfiguration(
    $name: String!
    $type_slug: String!
    $configuration_parameters: [ConfigurationParameterInput]
    $test_source_id: UUID
    $project_id: UUID!
    $has_pre_test: Boolean
    $has_post_test: Boolean
    $has_load_tests: Boolean
    $has_monitoring: Boolean
  ) {
    testrun_configuration_create(
      configuration_parameters: $configuration_parameters
      name: $name
      project_id: $project_id
      type_slug: $type_slug
      test_source_id: $test_source_id
      has_pre_test: $has_pre_test
      has_post_test: $has_post_test
      has_load_tests: $has_load_tests
      has_monitoring: $has_monitoring
    ) {
      returning {
        id
      }
    }
  }
`
export const EDIT_CONFIGURATION_MUTATION = gql`
  mutation editTestConfiguration(
    $id: UUID!
    $name: String!
    $type_slug: String!
    $configuration_parameters: [ConfigurationParameterInput]
    $test_source_id: UUID
    $has_pre_test: Boolean
    $has_post_test: Boolean
    $has_load_tests: Boolean
    $has_monitoring: Boolean
  ) {
    testrun_configuration_update(
      id: $id
      configuration_parameters: $configuration_parameters
      name: $name
      type_slug: $type_slug
      test_source_id: $test_source_id
      has_pre_test: $has_pre_test
      has_post_test: $has_post_test
      has_load_tests: $has_load_tests
      has_monitoring: $has_monitoring
    ) {
      returning {
        id
      }
    }
  }
`

export const EDIT_PERFORMED_CONFIGURATION_MUTATION = gql`
  mutation editPerformedTestConfiguration($id: UUID!, $name: String!) {
    testrun_configuration_update(id: $id, name: $name) {
      returning {
        id
      }
    }
  }
`

export const GET_TEST_SOURCES_FOR_PROJECT = gql`
  query getTestSourcesForProject($projectId: uuid) {
    testSources: test_source(where: { project_id: { _eq: $projectId } }) {
      id
      source_type
      repository {
        id
        name
        type_slug
      }
      test_creator {
        id
        name
        type_slug
      }
    }
  }
`

export const GET_CONFIGURATION_TYPES = gql`
  query getConfigurationTypesForSelector {
    configurationTypes: configuration_type {
      id
      name
      slug_name
    }
  }
`

export const GET_PARAMETERS = gql`
  query getParameters {
    parameters: parameter {
      id
      name
      param_name
      param_type
      default_value
      slug_name
      type_slug
    }
  }
`
