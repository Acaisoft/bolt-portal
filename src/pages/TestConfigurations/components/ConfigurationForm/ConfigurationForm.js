import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Field } from 'react-final-form'
import { Button, Grid, MenuItem, withStyles } from '@material-ui/core'
import { FormField } from '~containers'
import { TestConfigurationForm } from '~containers/forms'
import { ExpandablePanel, SectionHeader } from '~components'

import styles from './ConfigurationForm.styles'

export class ConfigurationForm extends Component {
  static propTypes = {
    onCancel: PropTypes.func,
    onSubmit: PropTypes.func,
    configurationId: PropTypes.string,
    projectId: PropTypes.string,
  }

  static defaultProps = {
    onCancel: () => {},
    onSubmit: () => {},
  }

  render() {
    const { classes, onCancel, onSubmit, configurationId, projectId } = this.props
    const mode = configurationId ? 'edit' : 'create'

    return (
      <TestConfigurationForm
        mode={mode}
        configurationId={configurationId}
        projectId={projectId}
        onSubmit={onSubmit}
      >
        {({ form, fields }) => (
          <form onSubmit={form.handleSubmit}>
            <SectionHeader
              title={mode === 'create' ? 'New Scenario' : 'Update Scenario'}
              marginBottom
            >
              <Button
                color="default"
                variant="text"
                disabled={form.isSubmitting}
                onClick={onCancel}
                className={classes.button}
              >
                Cancel
              </Button>
              <Button
                color="secondary"
                variant="contained"
                type="submit"
                disabled={!form.dirty || form.isSubmitting || form.invalid}
                className={classes.button}
              >
                {mode === 'create' ? 'Create' : 'Update'}
              </Button>
            </SectionHeader>
            <ExpandablePanel defaultExpanded title="1. Scenario">
              <Grid container spacing={32}>
                <Grid item xs={6}>
                  <FormField
                    name="name"
                    field={fields.name}
                    fullWidth
                    variant="filled"
                  />
                </Grid>
                <Grid item xs={6}>
                  <FormField
                    name="configuration_type"
                    field={fields.configuration_type}
                    variant="filled"
                    fullWidth
                  >
                    {fields.configuration_type.options.map(option => (
                      <MenuItem key={option.key} value={option.value}>
                        {option.label}
                      </MenuItem>
                    ))}
                  </FormField>
                </Grid>
              </Grid>
            </ExpandablePanel>

            <ExpandablePanel defaultExpanded title="2. Test Parameters">
              <Grid container spacing={32}>
                {Object.entries(fields.parameters.fields || [])
                  .filter(
                    ([name, options]) =>
                      options.group === form.values.configuration_type
                  )
                  .map(([id, options]) => (
                    <Grid key={id} item xs={6}>
                      <FormField
                        name={`parameters.${id}`}
                        field={options}
                        fullWidth
                        variant="filled"
                      />
                    </Grid>
                  ))}
              </Grid>
            </ExpandablePanel>

            <ExpandablePanel defaultExpanded title="3. Test Source">
              <Grid container spacing={32}>
                <Grid item xs={6}>
                  <FormField
                    name="test_source_type"
                    field={fields.test_source_type}
                    fullWidth
                    variant="filled"
                  >
                    {fields.test_source_type.options.map(option => (
                      <MenuItem key={option.key} value={option.value}>
                        {option.label}
                      </MenuItem>
                    ))}
                  </FormField>
                </Grid>
                <Grid item xs={6}>
                  <Field name="test_source_type" subscription={{ value: true }}>
                    {({ input: { value: selectedSourceType } }) => {
                      if (!selectedSourceType) {
                        return null
                      }

                      return (
                        <FormField
                          name={`test_source.${selectedSourceType}`}
                          field={fields.test_source.fields[selectedSourceType]}
                          fullWidth
                          variant="filled"
                        >
                          {fields.test_source.fields[selectedSourceType].options.map(
                            option => (
                              <MenuItem key={option.key} value={option.value}>
                                {option.label}
                              </MenuItem>
                            )
                          )}
                        </FormField>
                      )
                    }}
                  </Field>
                </Grid>
              </Grid>
            </ExpandablePanel>
          </form>
        )}
      </TestConfigurationForm>
    )
  }
}

export default withStyles(styles)(ConfigurationForm)
