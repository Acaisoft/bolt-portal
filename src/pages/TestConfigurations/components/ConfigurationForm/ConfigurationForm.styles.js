export default ({ spacing }) => ({
  button: {
    marginLeft: spacing.unit * 2,
  },
})
