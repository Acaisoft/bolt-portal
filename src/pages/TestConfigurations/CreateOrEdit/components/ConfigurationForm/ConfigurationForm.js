import React, { useCallback, useMemo } from 'react'
import PropTypes from 'prop-types'
import { useQuery } from 'react-apollo-hooks'

import { Field, Form } from 'react-final-form'
import { Button, Grid, MenuItem, withStyles, Typography } from '@material-ui/core'
import { FormField, CheckboxField } from '~containers'
import { ExpandablePanel, SectionHeader, Loader } from '~components'

import {
  makeEmptyInitialValues,
  validateForm,
  makeFlatValidationSchema,
} from '~utils/forms'
import { useMutationWithState } from '~hooks'

import {
  GET_CONFIGURATION,
  ADD_CONFIGURATION_MUTATION,
  EDIT_PERFORMED_CONFIGURATION_MUTATION,
  EDIT_CONFIGURATION_MUTATION,
} from './graphql'
import { useFormSchema, prepareInitialValues, preparePayload } from './formSchema'
import styles from './ConfigurationForm.styles'

export function ConfigurationForm({
  classes,
  mode,
  onCancel = () => {},
  onSubmit = () => {},
  configurationId,
  projectId,
}) {
  const { data: { configuration } = {}, loading: configurationLoading } = useQuery(
    GET_CONFIGURATION,
    {
      fetchPolicy: 'cache-and-network',
      variables: { configurationId },
      skip: mode !== 'edit',
    }
  )
  const isPerformed = Boolean(configuration && configuration.performed)

  const { fields, loading: fieldsLoading } = useFormSchema({ mode, projectId })

  const handleSubmit = useConfigurationSubmit({
    configurationId,
    projectId,
    mode,
    isPerformed,
    onSubmit,
  })
  const handleValidate = useCallback(
    values => validateForm(values, makeFlatValidationSchema(fields)),
    [fields]
  )

  const initialValues = useMemo(
    () => makeEmptyInitialValues(fields, prepareInitialValues(configuration)),
    [fields, configuration]
  )

  if (fieldsLoading || configurationLoading) {
    return <Loader loading />
  }

  return (
    <Form
      initialValues={initialValues}
      onSubmit={handleSubmit}
      validate={handleValidate}
      subscription={{ submitting: true, dirty: true, invalid: true }}
      keepDirtyOnReinitialize
    >
      {form => (
        <form onSubmit={form.handleSubmit}>
          <SectionHeader
            title={mode === 'create' ? 'New Scenario' : 'Update Scenario'}
            description={
              isPerformed
                ? 'We want to keep your data consistent and be able to prepare really precisely reports. This is why you canâ€™t change the parameters of already performed scenarios.'
                : null
            }
            alignItems="flex-start"
            marginBottom
          >
            <Button
              color="default"
              variant="text"
              disabled={form.isSubmitting}
              onClick={onCancel}
              className={classes.button}
            >
              Cancel
            </Button>
            <Button
              color="secondary"
              variant="contained"
              type="submit"
              disabled={!form.dirty || form.isSubmitting || form.invalid}
              className={classes.button}
            >
              {mode === 'create' ? 'Create' : 'Update'}
            </Button>
          </SectionHeader>
          <ExpandablePanel defaultExpanded title="Scenario">
            <Grid container spacing={32}>
              <Grid item xs={6}>
                <FormField
                  name="scenario_name"
                  field={fields.scenario_name}
                  fullWidth
                  variant="filled"
                />
              </Grid>
              <Grid item xs={6}>
                <FormField
                  name="configuration_type"
                  field={fields.configuration_type}
                  disabled={isPerformed}
                  variant="filled"
                  fullWidth
                >
                  {fields.configuration_type.options.map(option => (
                    <MenuItem key={option.key} value={option.value}>
                      {option.label}
                    </MenuItem>
                  ))}
                </FormField>
              </Grid>
            </Grid>
          </ExpandablePanel>

          <ExpandablePanel defaultExpanded title="Scenario Parts">
            <Grid container spacing={16}>
              {Object.entries(fields.scenario_parts.fields).map(([name, field]) => (
                <Grid item xs={12} md={6} key={name}>
                  <FormField
                    name={`scenario_parts.${name}`}
                    field={field}
                    disabled={isPerformed}
                    type="checkbox"
                    component={CheckboxField}
                  />
                </Grid>
              ))}
            </Grid>
          </ExpandablePanel>

          <Field name="configuration_type" subscription={{ value: true }}>
            {({ input: { value: configurationType } }) => (
              <React.Fragment>
                <Field
                  type="checkbox"
                  name="scenario_parts.has_load_tests"
                  subscription={{ value: true }}
                >
                  {({ input: { value: hasLoadTests } }) => (
                    <ExpandablePanel
                      key={`${configurationType}-step2`} // Re-render to reinitialize defaultExpanded
                      defaultExpanded={Boolean(configurationType)}
                      title="Test Parameters"
                    >
                      <Grid container spacing={32}>
                        {!configurationType || !hasLoadTests ? (
                          <Grid item xs={12}>
                            <Typography variant="body1">
                              Select test type first to see available parameters list
                            </Typography>
                          </Grid>
                        ) : (
                          Object.entries(fields.parameters.fields || [])
                            .filter(
                              ([name, options]) =>
                                options.group === configurationType
                            )
                            .map(([id, options]) => (
                              <Grid key={id} item xs={6}>
                                <FormField
                                  name={`parameters.${id}`}
                                  field={options}
                                  disabled={isPerformed}
                                  fullWidth
                                  variant="filled"
                                />
                              </Grid>
                            ))
                        )}
                      </Grid>
                    </ExpandablePanel>
                  )}
                </Field>

                <ExpandablePanel
                  key={`${configurationType}-step3`} // Re-render to reinitialize defaultExpanded
                  defaultExpanded={Boolean(configurationType)}
                  title="Test Source"
                >
                  <Grid container spacing={32}>
                    {!configurationType ? (
                      <Grid item xs={12}>
                        <Typography variant="body1">
                          Select test type first to see test source options
                        </Typography>
                      </Grid>
                    ) : (
                      <React.Fragment>
                        <Grid item xs={6}>
                          <FormField
                            name="test_source_type"
                            field={fields.test_source_type}
                            disabled={isPerformed}
                            fullWidth
                            variant="filled"
                          >
                            {fields.test_source_type.options.map(option => (
                              <MenuItem key={option.key} value={option.value}>
                                {option.label}
                              </MenuItem>
                            ))}
                          </FormField>
                        </Grid>
                        <Grid item xs={6}>
                          <Field
                            name="test_source_type"
                            subscription={{ value: true }}
                          >
                            {({ input: { value: selectedSourceType } }) => {
                              if (!selectedSourceType) {
                                return null
                              }

                              return (
                                <FormField
                                  name={`test_source.${selectedSourceType}`}
                                  field={
                                    fields.test_source.fields[selectedSourceType]
                                  }
                                  disabled={isPerformed}
                                  fullWidth
                                  variant="filled"
                                >
                                  {fields.test_source.fields[
                                    selectedSourceType
                                  ].options.map(option => (
                                    <MenuItem key={option.key} value={option.value}>
                                      {option.label}
                                    </MenuItem>
                                  ))}
                                </FormField>
                              )
                            }}
                          </Field>
                        </Grid>
                      </React.Fragment>
                    )}
                  </Grid>
                </ExpandablePanel>
              </React.Fragment>
            )}
          </Field>
        </form>
      )}
    </Form>
  )
}
ConfigurationForm.propTypes = {
  mode: PropTypes.oneOf(['create', 'edit']),
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  configurationId: PropTypes.string,
  projectId: PropTypes.string,
}

function useConfigurationSubmit({
  mode,
  isPerformed,
  configurationId,
  projectId,
  onSubmit,
}) {
  const { mutation: submitMutation } = useMutationWithState(
    mode === 'create'
      ? ADD_CONFIGURATION_MUTATION
      : isPerformed
      ? EDIT_PERFORMED_CONFIGURATION_MUTATION
      : EDIT_CONFIGURATION_MUTATION,
    {
      refetchQueries: [
        'getTestConfigurations',
        mode === 'edit' ? 'getConfiguration' : '',
      ],
    }
  )

  const handleSubmit = useCallback(
    async values => {
      const dbValues = preparePayload(values)

      let variables
      if (isPerformed) {
        variables = { id: configurationId, name: dbValues.name }
      } else {
        variables = dbValues

        if (mode === 'create') {
          variables.project_id = projectId
        } else {
          variables.id = configurationId
        }
      }

      const { errorMessage } = await submitMutation({ variables })

      onSubmit({ values, errorMessage })
    },
    [submitMutation, isPerformed, mode, configurationId, projectId, onSubmit]
  )

  return handleSubmit
}
export default withStyles(styles)(ConfigurationForm)
