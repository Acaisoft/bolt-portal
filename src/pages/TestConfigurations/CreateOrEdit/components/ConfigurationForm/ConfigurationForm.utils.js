import { useCallback } from 'react'
import { useMutationWithState } from 'hooks'
import { preparePayload } from './formSchema'
import { ADD_CONFIGURATION_MUTATION, EDIT_CONFIGURATION_MUTATION } from './graphql'

export const testsPerformedMessage =
  'You cannot change test source - a test has been performed'

export function useConfigurationSubmit({
  mode,
  configurationId,
  projectId,
  onSubmit,
}) {
  const { mutation: submitMutation } = useMutationWithState(
    mode === 'create' ? ADD_CONFIGURATION_MUTATION : EDIT_CONFIGURATION_MUTATION,
    {
      refetchQueries: [
        'getTestConfigurations',
        mode === 'edit' ? 'getConfiguration' : '',
      ],
    }
  )

  const handleSubmit = useCallback(
    async values => {
      const variables = preparePayload(values, {
        mode,
        configurationId,
        projectId,
      })
      const { errorMessage } = await submitMutation({ variables })
      onSubmit({ values, errorMessage })
    },
    [submitMutation, mode, configurationId, projectId, onSubmit]
  )

  return handleSubmit
}
