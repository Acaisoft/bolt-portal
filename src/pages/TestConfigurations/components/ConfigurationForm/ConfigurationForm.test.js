import React from 'react'
import { shallow } from 'enzyme'

import { ConfigurationForm } from './ConfigurationForm'

const initConfigurationForm = overrides => {
  const mockProps = {
    configurationId: 'testConfigurationId',
    projectId: 'testProjectId',
  }
  const mockMethods = {
    onCancel: jest.fn(),
    onSubmit: jest.fn(),
  }
  const wrapper = shallow(<ConfigurationForm {...mockProps} {...overrides} />)
  return { mockProps, wrapper }
}

describe('global: ConfigurationForm', () => {
  it('renders without crashing', () => {
    const { wrapper } = initConfigurationForm()
    expect(wrapper).toBeTruthy()
  })

  it('should render as expected', () => {
    const { wrapper } = initConfigurationForm({})
    expect(wrapper).toMatchSnapshot()
  })
})
