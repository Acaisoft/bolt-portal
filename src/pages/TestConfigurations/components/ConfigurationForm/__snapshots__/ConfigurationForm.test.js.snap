// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`global: ConfigurationForm should render as expected 1`] = `
<Apollo(Apollo(Apollo(Apollo(TestConfigurationForm))))
  configurationId="testConfigurationId"
  mode="edit"
  onSubmit={[Function]}
  projectId="testProjectId"
>
  <Component />
</Apollo(Apollo(Apollo(Apollo(TestConfigurationForm))))>
`;
